# Chapter 7 Exercise Solutions

---

#### Exercise 7.1

$$ \begin{align*}
    4x_1 +x_2 -x_3 +x_4 &=14,\\
    x_1 +4x_2 -x_3 -x_4 &=10,\\
    -x_1 -x_2 +5x_3 +x_4 &=-15,\\
    x_1 -x_2 +x_3 +3x_4 &=3.
\end{align*} $$

The Jacobi method for this system is

$$ \begin{align*}
  x_1^{(k+1)} &= \frac{1}{4}(14 - x_2^{(k)} + x_3^{(k)} - x_4^{(k)}), \\
  x_2^{(k+1)} &= \frac{1}{4}(10 - x_1^{(k)} + x_3^{(k)} + x_4^{(k)}), \\
  x_3^{(k+1)} &= \frac{1}{5}(-15 + x_1^{(k)} + x_2^{(k)} - x_4^{(k)}), \\
  x_4^{(k+1)} &= \frac{1}{3}(3 - x_1^{(k)} + x_2^{(k)} - x_3^{(k)}).
\end{align*} $$

First iteration:

$$ \begin{align*}
  x_1^{(1)} &= \frac{1}{4}(14 - 0 + 0 - 0) = 3.5, \\
  x_2^{(1)} &= \frac{1}{4}(10 - 0 + 0 + 0) = 2.5, \\
  x_3^{(1)} &= \frac{1}{5}(-15 + 0 + 0 - 0) = -3, \\
  x_4^{(1)} &= \frac{1}{3}(3 - 0 + 0 - 0) = 1.
\end{align*} $$

Second iteration:

$$ \begin{align*}
  x_1^{(2)} &= \frac{1}{4}(14 - 2.5 - 3 - 1) = 1.875, \\
  x_2^{(2)} &= \frac{1}{4}(10 - 3.5 - 3 + 1) = 1.125, \\
  x_3^{(2)} &= \frac{1}{5}(-15 + 3.5 + 2.5 - 1) = -2, \\
  x_4^{(2)} &= \frac{1}{3}(3 - 3.5 + 2.5 + 3) = 1.6667.
\end{align*} $$

---

#### Exercise 7.2

The Gauss-Seidel method for this system is

$$ \begin{align*}
  x_1^{(k+1)} &= \frac{1}{4}(14 - x_2^{(k)} + x_3^{(k)} - x_4^{(k)}), \\
  x_2^{(k+1)} &= \frac{1}{4}(10 - x_1^{(k+1)} + x_3^{(k)} + x_4^{(k)}), \\
  x_3^{(k+1)} &= \frac{1}{5}(-15 + x_1^{(k+1)} + x_2^{(k+1)} - x_4^{(k)}), \\
  x_4^{(k+1)} &= \frac{1}{3}(3 - x_1^{(k+1)} + x_2^{(k+1)} - x_3^{(k+1)}).
\end{align*} $$

First iteration:

$$ \begin{align*}
  x_1^{(1)} &= \frac{1}{4}(14 - 0 + 0 - 0) = 3.5, \\
  x_2^{(1)} &= \frac{1}{4}(10 - 3.5 + 0 + 0) = 1.625, \\
  x_3^{(1)} &= \frac{1}{5}(-15 + 3.5 + 1.625 - 0) = -1.975, \\
  x_4^{(1)} &= \frac{1}{3}(3 - 3.5 + 1.625 + 1.975) = 1.0333.
\end{align*} $$

Second iteration:

$$ \begin{align*}
  x_1^{(2)} &= \frac{1}{4}(14 - 1.625 - 1.975 - 1.0333) = 2.3417, \\
  x_2^{(2)} &= \frac{1}{4}(10 - 2.3417 - 1.975 + 1.0333) = 1.6792, \\
  x_3^{(2)} &= \frac{1}{5}(-15 + 2.3417 + 1.6792 - 1.0333) = -2.4025, \\
  x_4^{(2)} &= \frac{1}{3}(3 - 2.3417 + 1.6792 + 2.4025) = 1.58.
\end{align*} $$

---

#### Exercise 7.4

Checking that the coefficient matrix has all real eigenvalues.

In [71]:
import numpy as np

A = np.array([[ 4, 1, -1, 1 ], 
              [1, 4, -1, -1], 
              [-1, -1, 5, 1], 
              [1, -1, 1, 3]])

print("eigenvalues of A =", np.linalg.eig(A)[0])

eigenvalues of A = [1.41126245 6.58873755 4.54629468 3.45370532]


All eigenvalues of $A$ are real so we can use

$$ \omega = 1 + \left( \frac{\rho(T_J)}{1 + \sqrt{1 - \rho(T_J)^2}} \right)^2. $$

The iteration matrix for the Jacobi method is

$$ T_J = D^{-1} (L + U)$$

In [59]:
L = np.tril(A, -1)
U = np.triu(A, 1)
D = A - L - U

TJ = -np.dot(np.linalg.inv(D), L + U)
print("TJ = \n", TJ)

TJ = 
 [[-0.         -0.25        0.25       -0.25      ]
 [-0.25       -0.          0.25        0.25      ]
 [ 0.2         0.2        -0.         -0.2       ]
 [-0.33333333  0.33333333 -0.33333333 -0.        ]]


So $T_J = \begin{pmatrix} 0 & -\frac{1}{4} & \frac{1}{4} & -\frac{1}{4} \\ -\frac{1}{4} & 0 & \frac{1}{4} & \frac{1}{4} \\ \frac{1}{5} & \frac{1}{5} & 0 & -\frac{1}{5} \\ -\frac{1}{3} & \frac{1}{3} & -\frac{1}{3} & 0 \end{pmatrix}$.

Calculating $\rho(T_J)$

In [60]:
rhoTJ = max(abs(np.linalg.eig(TJ)[0]))
print("rhoTJ = ", rhoTJ)

rhoTJ =  0.6054108391837034


So $\rho(T_J) = 0.6054$, therefore
$$ \omega = 1 + \left( \frac{0.6054}{1 + \sqrt{1-0.6054^2}} \right)^2 = 1.1136.$$
The SOR method for this system is

$$ \begin{align*}
  x_1^{(k+1)} &= (1 - \omega)x_1^{(k)} + \frac{\omega}{4}(14 - x_2^{(k)} + x_3^{(k)} - x_4^{(k)}), \\
  x_2^{(k+1)} &= (1 - \omega)x_2^{(k)} + \frac{\omega}{4}(10 - x_1^{(k+1)} + x_3^{(k)} + x_4^{(k)}), \\
  x_3^{(k+1)} &= (1 - \omega)x_3^{(k)} + \frac{\omega}{5}(-15 + x_1^{(k+1)} + x_2^{(k+1)} - x_4^{(k)}), \\
  x_4^{(k+1)} &= (1 - \omega)x_4^{(k)} + \frac{\omega}{3}(3 - x_1^{(k+1)} + x_2^{(k+1)} - x_3^{(k+1)}).
\end{align*} $$

First iteration:

$$ \begin{align*}
  x_1^{(1)} &= (1 - 1.1136)(0) + \frac{1.1136}{4}(14 - 0 + 0 - 0) = 3.8977, \\
  x_2^{(1)} &= (1 - 1.1136)(0) + \frac{1.1136}{4}(10 - 3.8977 + 0 + 0) = 1.6989, \\
  x_3^{(1)} &= (1 - 1.1136)(0) + \frac{1.1136}{5}(-15 + 3.8977 + 1.6989 - 0) = -2.0944, \\
  x_4^{(1)} &= (1 - 1.1136)(0) + \frac{1.1136}{3}(3 - 3.8977 + 1.6989 + 2.0944) = 1.0749. 
\end{align*} $$

Second iteration:

$$ \begin{align*}
  x_1^{(2)} &= (1 - 1.1136)(3.8977) + \frac{1.1136}{4}(14 - 1.6989 - 2.0944 + 1.0749) = 2.0994, \\
  x_2^{(2)} &= (1 - 1.1136)(1.6989) + \frac{1.1136}{4}(10 - 2.0994 - 2.0994 - 1.0749) = 1.7227, \\
  x_3^{(2)} &= (1 - 1.1136)(-2.0944) + \frac{1.1136}{5}(-15 + 2.0994 + 1.7227 - 1.0749) = -2.4910, \\
  x_4^{(2)} &= (1 - 1.1136)(1.0749) + \frac{1.1136}{3}(3 - 2.0994 + 1.7227 + 2.4910) = 1.7763.
\end{align*} $$

---

#### Exercise 7.4

The iteration matrices for the Gauss-Seidel and SOR methods are

$$ \begin{align*}
  T_{GS} &= -(L+D)^{-1} U, \\
  T_{SOR} &= (D+\omega L)^{-1} ((1-\omega )D-\omega U).
\end{align*} $$

In [62]:
TGS = -np.dot(np.linalg.inv(L + D), U)
TSOR = np.dot(np.linalg.inv(D + omega * L), (1 - omega) * D - omega * U)

print("TGS = \n", TGS)
print("\nTSOR = \n", TSOR)

TGS = 
 [[-0.         -0.25        0.25       -0.25      ]
 [-0.          0.0625      0.1875      0.3125    ]
 [-0.         -0.0375      0.0875     -0.1875    ]
 [-0.          0.11666667 -0.05        0.25      ]]

TSOR = 
 [[-0.11363961 -0.2784099   0.2784099  -0.2784099 ]
 [ 0.03163839 -0.03612754  0.20089783  0.35592198]
 [-0.01826396 -0.07005627 -0.00688439 -0.20546382]
 [ 0.06070894  0.11594423 -0.02621793  0.19810364]]


In [63]:
rhoTGS = max(abs(np.linalg.eig(TGS)[0]))
rhoTSOR = max(abs(np.linalg.eig(TSOR)[0]))

print(f"rho(TGS) = {rhoTGS:0.4f}")
print(f"rho(TSOR) = {rhoTSOR:0.4f}")

rho(TGS) = 0.3553
rho(TSOR) = 0.1968


The spectral radii of the iteration matrices for the three methods are

$$ \begin{align*}
  \rho(T_J) &= 0.6054, \\
  \rho(T_{GS}) &= 0.3553, \\
  \rho(T_{SOR}) &= 0.1968.
\end{align*} $$

So the SOR method should converge the fastest.

---

#### Exercise 7.5

In [67]:
def jacobi(A, b, tol=1e-6):
    n = len(b)
    x = np.zeros(n)
    maxiter = 100
    for k in range(maxiter):
        xold = np.copy(x)
        for i in range(n):
            sum_ = 0
            for j in range(n):
                if i != j:
                    sum_ += A[i,j] * xold[j]
        
            x[i] = (b[i] - sum_) / A[i,i]
            
        r = b - np.dot(A, x)

        if max(abs(r)) < tol:
            break
    
    return x, k


def gauss_seidel(A, b, tol=1e-6):
    n = len(b)
    x = np.zeros(n)
    maxiter = 100
    for k in range(maxiter):
        for i in range(n):
            sum_ = 0
            for j in range(n):
                if i != j:
                    sum_ += A[i,j] * x[j]
        
            x[i] = (b[i] - sum_) / A[i,i]
            
        r = b - np.dot(A, x)

        if max(abs(r)) < tol:
            break
    
    return x, k


def sor(A, b, omega, tol):
    n = len(b)
    x = np.zeros(n)
    maxiter = 100
    for k in range(maxiter):
        for i in range(n):
            sum_ = 0
            for j in range(n):
                if i != j:
                    sum_ += A[i,j] * x[j]
        
            x[i] = (1 - omega) * x[i] + omega * (b[i] - sum_) / A[i,i]
            
        r = b - np.dot(A, x)
        if max(abs(r)) < tol:
            break
    
    return x, k


 # Define system
A = np.array([[ 4, 1, -1, 1 ], 
              [1, 4, -1, -1], 
              [-1, -1, 5, 1], 
              [1, -1, 1, 3]])
b = np.array([14, 10, -15, 3])

# Solve system
tol = 1e-6
x, jacobi_iterations = jacobi(A, b, tol)
x, GS_iterations = gauss_seidel(A, b, tol)
x, SOR_iterations = sor(A, b, omega, tol)

print(f"x = {x}")
print(f"\nThe Jacobi method took {jacobi_iterations} iterations to converge")
print(f"The Gauss-Seidel method took {GS_iterations} iterations to converge")
print(f"The SOR method took {SOR_iterations} iterations to converge")

x = [ 1.93150692  1.82191783 -2.61643835  1.83561641]

The Jacobi method took 28 iterations to converge
The Gauss-Seidel method took 15 iterations to converge
The SOR method took 10 iterations to converge


---

#### Exercise 7.6

The Jacobi iteration matrix for this system is

$$ \begin{align*}
  T_J &= -D^{-1}(L + U) 
  = \begin{pmatrix} 2 & 0 \\ 0 & 2 \end{pmatrix}^{-1} \begin{pmatrix} 0 & 1 \\ \alpha & 0 \end{pmatrix}
  = \begin{pmatrix} \frac{1}{2} & 0 \\ 0 & \frac{1}{2} \end{pmatrix} \begin{pmatrix} 0 & 1 \\ \alpha & 0 \end{pmatrix}
  = \begin{pmatrix}  0 & \frac{1}{2} \\ \frac{1}{2}\alpha & 0 \end{pmatrix}.
\end{align*} $$

Calculating the eigenvalues of $T_J$

$$ \begin{align*}
  0 &= \begin{vmatrix} -\lambda & \frac{1}{2} \\ \frac{1}{2}\alpha & -\lambda \end{vmatrix} 
  = \lambda^2 - \frac{1}{4} \alpha
\end{align*} $$

therefore $\lambda = \pm\frac{1}{2}\sqrt{\alpha}$. For convergence $\rho(T_J) <1$ so $\frac{1}{2}\sqrt{\alpha} < 1$ therefore $\alpha < \sqrt{2}$.